# Lab 5 Exercise 1

# Setup

In [1]:
import warnings

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

warnings.filterwarnings('ignore')

from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, r2_score

# Importing Dataset

In [2]:
df = pd.read_csv('weatherAUS.csv')
df.head()

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0  2008-12-01   Albury     13.4     22.9       0.6          NaN       NaN   
1  2008-12-02   Albury      7.4     25.1       0.0          NaN       NaN   
2  2008-12-03   Albury     12.9     25.7       0.0          NaN       NaN   
3  2008-12-04   Albury      9.2     28.0       0.0          NaN       NaN   
4  2008-12-05   Albury     17.5     32.3       1.0          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
0           W           44.0          W  ...        71.0         22.0   
1         WNW           44.0        NNW  ...        44.0         25.0   
2         WSW           46.0          W  ...        38.0         30.0   
3          NE           24.0         SE  ...        45.0         16.0   
4           W           41.0        ENE  ...        82.0         33.0   

   Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
0       1007.7       1007.1       8.0       NaN     16.9     21.8         No   
1       1010.6       1007.8       NaN       NaN     17.2     24.3         No   
2       1007.6       1008.7       NaN       2.0     21.0     23.2         No   
3       1017.6       1012.8       NaN       NaN     18.1     26.5         No   
4       1010.8       1006.0       7.0       8.0     17.8     29.7         No   

   RainTomorrow  
0            No  
1            No  
2            No  
3            No  
4            No  

[5 rows x 23 columns]

# Checking for NaN values

In [3]:
df.isna().sum()

Date                 0
Location             0
MinTemp           1485
MaxTemp           1261
Rainfall          3261
Evaporation      62790
Sunshine         69835
WindGustDir      10326
WindGustSpeed    10263
WindDir9am       10566
WindDir3pm        4228
WindSpeed9am      1767
WindSpeed3pm      3062
Humidity9am       2654
Humidity3pm       4507
Pressure9am      15065
Pressure3pm      15028
Cloud9am         55888
Cloud3pm         59358
Temp9am           1767
Temp3pm           3609
RainToday         3261
RainTomorrow      3267
dtype: int64

# Information about Dataset

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Date           145460 non-null  object 
 1   Location       145460 non-null  object 
 2   MinTemp        143975 non-null  float64
 3   MaxTemp        144199 non-null  float64
 4   Rainfall       142199 non-null  float64
 5   Evaporation    82670 non-null   float64
 6   Sunshine       75625 non-null   float64
 7   WindGustDir    135134 non-null  object 
 8   WindGustSpeed  135197 non-null  float64
 9   WindDir9am     134894 non-null  object 
 10  WindDir3pm     141232 non-null  object 
 11  WindSpeed9am   143693 non-null  float64
 12  WindSpeed3pm   142398 non-null  float64
 13  Humidity9am    142806 non-null  float64
 14  Humidity3pm    140953 non-null  float64
 15  Pressure9am    130395 non-null  float64
 16  Pressure3pm    130432 non-null  float64
 17  Cloud9am       89572 non-null

### Dropping Column which is not Useful

In [5]:
df.drop('Date', axis=1, inplace=True)

# Dropping NaN values

In [6]:
df.dropna(inplace=True)

In [7]:
df.shape

(56420, 22)

# Splitting into X and y dataset

In [8]:
target = df.pop('RainTomorrow')
df.head()

Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine WindGustDir  \
6049    Cobar     17.9     35.2       0.0         12.0      12.3         SSW   
6050    Cobar     18.4     28.9       0.0         14.8      13.0           S   
6052    Cobar     19.4     37.6       0.0         10.8      10.6         NNE   
6053    Cobar     21.9     38.4       0.0         11.4      12.2         WNW   
6054    Cobar     24.2     41.0       0.0         11.2       8.4         WNW   

      WindGustSpeed WindDir9am WindDir3pm  ...  WindSpeed3pm  Humidity9am  \
6049           48.0        ENE         SW  ...          20.0         20.0   
6050           37.0        SSE        SSE  ...          19.0         30.0   
6052           46.0        NNE        NNW  ...          15.0         42.0   
6053           31.0        WNW        WSW  ...           6.0         37.0   
6054           35.0         NW        WNW  ...          13.0         19.0   

      Humidity3pm  Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  \
6049         13.0       1006.3       1004.4       2.0       5.0     26.6   
6050          8.0       1012.9       1012.1       1.0       1.0     20.3   
6052         22.0       1012.3       1009.2       1.0       6.0     28.7   
6053         22.0       1012.7       1009.1       1.0       5.0     29.1   
6054         15.0       1010.7       1007.4       1.0       6.0     33.6   

      Temp3pm  RainToday  
6049     33.4         No  
6050     27.0         No  
6052     34.9         No  
6053     35.6         No  
6054     37.6         No  

[5 rows x 21 columns]

# Mapping values of y to Numeric Values

In [9]:
mapping = {'No': 0, 'Yes': 1}
target = target.map(mapping)

In [10]:
df.dtypes

Location          object
MinTemp          float64
MaxTemp          float64
Rainfall         float64
Evaporation      float64
Sunshine         float64
WindGustDir       object
WindGustSpeed    float64
WindDir9am        object
WindDir3pm        object
WindSpeed9am     float64
WindSpeed3pm     float64
Humidity9am      float64
Humidity3pm      float64
Pressure9am      float64
Pressure3pm      float64
Cloud9am         float64
Cloud3pm         float64
Temp9am          float64
Temp3pm          float64
RainToday         object
dtype: object

# Splitting into Numeric and Categorical Columns

In [11]:
# for numerical_cols
numerical_cols   = [cname for cname in df.columns if df[cname].dtype in ['int64', 'float64']]\
# for categorical_cols
categorical_cols = [cname for cname in df.columns if df[cname].dtype == 'object']

In [12]:
df_num = df[numerical_cols]
df_cat = df[categorical_cols]

# Applying VIF on the Dataset

In [13]:
# Using Linear Regression and calculating the r2 scores
lr = LinearRegression()
vif = {}
for col in numerical_cols:
    all_cols = list(numerical_cols)
    all_cols.remove(col)
    X = df_num[all_cols]
    y = df_num[col]
    lr.fit(X, y)
    y_pred = lr.predict(X)
    score = r2_score(y, y_pred)
    vif[col] = 1 / (1 - score)
    
vif

{'MinTemp': 10.770219691409892,
 'MaxTemp': 46.78406795924843,
 'Rainfall': 1.1779953082896788,
 'Evaporation': 2.22481528634829,
 'Sunshine': 3.307730755606259,
 'WindGustSpeed': 2.878864348472314,
 'WindSpeed9am': 1.8564893477202455,
 'WindSpeed3pm': 2.1463942950501775,
 'Humidity9am': 4.417547175413456,
 'Humidity3pm': 6.803333309913033,
 'Pressure9am': 19.87624174315856,
 'Pressure3pm': 19.86913652734502,
 'Cloud9am': 2.238263376838908,
 'Cloud3pm': 2.2774283640218864,
 'Temp9am': 24.36887019910182,
 'Temp3pm': 56.08268800345098}

# Dropping Columns with VIF value less than 5

In [14]:
drop_cols = [key for key in vif.keys() if vif[key] > 5]
drop_cols

['MinTemp',
 'MaxTemp',
 'Humidity3pm',
 'Pressure9am',
 'Pressure3pm',
 'Temp9am',
 'Temp3pm']

In [15]:
df_num.drop(drop_cols, axis=1, inplace=True)

# Correlation Matrix

In [16]:
df_num.corr()

Rainfall  Evaporation  Sunshine  WindGustSpeed  WindSpeed9am  \
Rainfall       1.000000    -0.077239 -0.246379       0.106308      0.050584   
Evaporation   -0.077239     1.000000  0.369250       0.209566      0.193154   
Sunshine      -0.246379     0.369250  1.000000      -0.052422     -0.013842   
WindGustSpeed  0.106308     0.209566 -0.052422       1.000000      0.608852   
WindSpeed9am   0.050584     0.193154 -0.013842       0.608852      1.000000   
WindSpeed3pm   0.044112     0.124345  0.029200       0.685236      0.502226   
Humidity9am    0.263625    -0.554232 -0.500343      -0.193410     -0.236795   
Cloud9am       0.217169    -0.199809 -0.677939       0.088129      0.034908   
Cloud3pm       0.191433    -0.202366 -0.702022       0.131590      0.062507   

               WindSpeed3pm  Humidity9am  Cloud9am  Cloud3pm  
Rainfall           0.044112     0.263625  0.217169  0.191433  
Evaporation        0.124345    -0.554232 -0.199809 -0.202366  
Sunshine           0.029200    -0.500343 -0.677939 -0.702022  
WindGustSpeed      0.685236    -0.193410  0.088129  0.131590  
WindSpeed9am       0.502226    -0.236795  0.034908  0.062507  
WindSpeed3pm       1.000000    -0.100626  0.068224  0.041475  
Humidity9am       -0.100626     1.000000  0.438962  0.348707  
Cloud9am           0.068224     0.438962  1.000000  0.614380  
Cloud3pm           0.041475     0.348707  0.614380  1.000000

**Conclusion from the Above Correlation Matrix**
* Considering variables that have correlation more than 0.6
* WindGustSpeed, WindSpeed3pm
* WindGustSpeed, WindSpeed9am
* Cloud9am, Cloud3pm

# Removing Columns with Mulicollinearity

In [17]:
numerical_cols = list(numerical_cols)
numerical_cols.remove('Cloud9am')
numerical_cols.remove('Cloud3pm')
numerical_cols.remove('WindGustSpeed')
numerical_cols.remove('WindSpeed3pm')
numerical_cols.remove('WindSpeed9am')

# Making new Columns with Related Columns

In [18]:
df_num = df[numerical_cols]
df_num['Cloud9am - Cloud3pm'] = df['Cloud9am'] - df['Cloud3pm']
df_num['WindGustSpeed - WindSpeed3pm'] = df['WindGustSpeed'] - df['WindSpeed3pm']
df_num['WindGustSpeed - WindSpeed9am'] = df['WindGustSpeed'] - df['WindSpeed9am']
df_num.head()

MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  Humidity9am  \
6049     17.9     35.2       0.0         12.0      12.3         20.0   
6050     18.4     28.9       0.0         14.8      13.0         30.0   
6052     19.4     37.6       0.0         10.8      10.6         42.0   
6053     21.9     38.4       0.0         11.4      12.2         37.0   
6054     24.2     41.0       0.0         11.2       8.4         19.0   

      Humidity3pm  Pressure9am  Pressure3pm  Temp9am  Temp3pm  \
6049         13.0       1006.3       1004.4     26.6     33.4   
6050          8.0       1012.9       1012.1     20.3     27.0   
6052         22.0       1012.3       1009.2     28.7     34.9   
6053         22.0       1012.7       1009.1     29.1     35.6   
6054         15.0       1010.7       1007.4     33.6     37.6   

      Cloud9am - Cloud3pm  WindGustSpeed - WindSpeed3pm  \
6049                 -3.0                          28.0   
6050                  0.0                          18.0   
6052                 -5.0                          31.0   
6053                 -4.0                          25.0   
6054                 -5.0                          22.0   

      WindGustSpeed - WindSpeed9am  
6049                          42.0  
6050                          18.0  
6052                          16.0  
6053                          25.0  
6054                          18.0

# Merging both the Numerical and Categorical Columns

In [19]:
temp_df = pd.merge(df_num, df_cat, left_index = True, right_index = True)
temp_df

MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  Humidity9am  \
6049       17.9     35.2       0.0         12.0      12.3         20.0   
6050       18.4     28.9       0.0         14.8      13.0         30.0   
6052       19.4     37.6       0.0         10.8      10.6         42.0   
6053       21.9     38.4       0.0         11.4      12.2         37.0   
6054       24.2     41.0       0.0         11.2       8.4         19.0   
...         ...      ...       ...          ...       ...          ...   
142298     19.3     33.4       0.0          6.0      11.0         63.0   
142299     21.2     32.6       0.0          7.6       8.6         56.0   
142300     20.7     32.8       0.0          5.6      11.0         46.0   
142301     19.5     31.8       0.0          6.2      10.6         62.0   
142302     20.2     31.7       0.0          5.6      10.7         73.0   

        Humidity3pm  Pressure9am  Pressure3pm  Temp9am  Temp3pm  \
6049           13.0       1006.3       1004.4     26.6     33.4   
6050            8.0       1012.9       1012.1     20.3     27.0   
6052           22.0       1012.3       1009.2     28.7     34.9   
6053           22.0       1012.7       1009.1     29.1     35.6   
6054           15.0       1010.7       1007.4     33.6     37.6   
...             ...          ...          ...      ...      ...   
142298         32.0       1013.9       1010.5     24.5     32.3   
142299         28.0       1014.6       1011.2     24.8     32.0   
142300         23.0       1015.3       1011.8     24.8     32.1   
142301         58.0       1014.9       1010.7     24.8     29.2   
142302         32.0       1013.9       1009.7     25.4     31.0   

        Cloud9am - Cloud3pm  WindGustSpeed - WindSpeed3pm  \
6049                   -3.0                          28.0   
6050                    0.0                          18.0   
6052                   -5.0                          31.0   
6053                   -4.0                          25.0   
6054                   -5.0                          22.0   
...                     ...                           ...   
142298                 -1.0                          15.0   
142299                  7.0                          26.0   
142300                  0.0                          22.0   
142301                  0.0                           9.0   
142302                  1.0                          23.0   

        WindGustSpeed - WindSpeed9am Location WindGustDir WindDir9am  \
6049                            42.0    Cobar         SSW        ENE   
6050                            18.0    Cobar           S        SSE   
6052                            16.0    Cobar         NNE        NNE   
6053                            25.0    Cobar         WNW        WNW   
6054                            18.0    Cobar         WNW         NW   
...                              ...      ...         ...        ...   
142298                          26.0   Darwin         ENE         SE   
142299                          24.0   Darwin           E         SE   
142300                          16.0   Darwin           E          E   
142301                          17.0   Darwin         ESE         SE   
142302                          15.0   Darwin         ENE        ENE   

       WindDir3pm RainToday  
6049           SW        No  
6050          SSE        No  
6052          NNW        No  
6053          WSW        No  
6054          WNW        No  
...           ...       ...  
142298         NE        No  
142299         SE        No  
142300          W        No  
142301        NNW        No  
142302        NNW        No  

[56420 rows x 19 columns]

# One Hot Encoding the Categorical Columns

In [20]:
df_new = pd.get_dummies(temp_df)
df_new.head()

MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  Humidity9am  \
6049     17.9     35.2       0.0         12.0      12.3         20.0   
6050     18.4     28.9       0.0         14.8      13.0         30.0   
6052     19.4     37.6       0.0         10.8      10.6         42.0   
6053     21.9     38.4       0.0         11.4      12.2         37.0   
6054     24.2     41.0       0.0         11.2       8.4         19.0   

      Humidity3pm  Pressure9am  Pressure3pm  Temp9am  ...  WindDir3pm_S  \
6049         13.0       1006.3       1004.4     26.6  ...             0   
6050          8.0       1012.9       1012.1     20.3  ...             0   
6052         22.0       1012.3       1009.2     28.7  ...             0   
6053         22.0       1012.7       1009.1     29.1  ...             0   
6054         15.0       1010.7       1007.4     33.6  ...             0   

      WindDir3pm_SE  WindDir3pm_SSE  WindDir3pm_SSW  WindDir3pm_SW  \
6049              0               0               0              1   
6050              0               1               0              0   
6052              0               0               0              0   
6053              0               0               0              0   
6054              0               0               0              0   

      WindDir3pm_W  WindDir3pm_WNW  WindDir3pm_WSW  RainToday_No  \
6049             0               0               0             1   
6050             0               0               0             1   
6052             0               0               0             1   
6053             0               0               1             1   
6054             0               1               0             1   

      RainToday_Yes  
6049              0  
6050              0  
6052              0  
6053              0  
6054              0  

[5 rows x 90 columns]

# Converting all Values to Same Datatype

In [21]:
df_new = df_new.astype('float64')
target = target.astype('float64')

# Splitting Dataset into Train and Test Set

In [22]:
X_train, X_test, y_train, y_test = train_test_split(df_new, target, test_size = 0.3, random_state = 42)

# Applying Logistic Regression on The Train Set

In [23]:
model = LogisticRegression(random_state=42)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
y_pred

array([0., 0., 0., ..., 0., 0., 0.])

# Calculating the Accuracy Score

In [24]:
accuracy_score(y_test, y_pred)

0.8539524991137895

# Conclusion

* The Accuracy Score using Logistic Regression is ~85%